### PR TITLE
Implement retry attempt logic

### PR DESCRIPTION
**Pull request by carlosalberto**
The Java tracer, unlike the Go or Python tracers, drops Spans upon failed report requests. This PR addresses this in the follow fashion:

* The selection and logic is essentially the same as the Go tracer, which does a best-effort to keep Spans from a previous report, depending on the max buffer size.
* The minimum poll period was changed to 500ms (from 40ms), in order to prevent busy CPU cycles even in the presence of exponential retry logic. 500ms is also used by the Go tracer.
* By default we keep the legacy behavior of dropping Spans upon failure - done to potentially break existing users. Otherwise, to enable this, `withDropSpansOnFailure(false)` has to be set.
* Otherwise, include the error message when the http transport fails, as status code is barely useful.

### CODE REVIEW

**Review by tedsuo — APPROVED** [review]
LGTM

File: common/src/main/java/com/lightstep/tracer/shared/ClientMetrics.java [hunk 1]
```diff
@@ -4,21 +4,21 @@
 import com.lightstep.tracer.grpc.MetricsSample;
 
 import java.util.ArrayList;
-import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicInteger;
```
Comment (codeboten) [inline common/src/main/java/com/lightstep/tracer/shared/ClientMetrics.java:7]: Why the change to integer?

**Review by tylerbenson — APPROVED** [review]
Looks ok to me.

File: common/src/main/java/com/lightstep/tracer/shared/ClientMetrics.java (hunk 1 above)
Comment (carlosalberto) [reply common/src/main/java/com/lightstep/tracer/shared/ClientMetrics.java:7]: We hopefully don't drop as many Spans that we require a `long` :) But more importantly, this was creating warnings all over the place as we had been bad at keeping uniformity for this very metric (in some places we used int, in some other ones long, etc).
